### ISSUE TITLE
WeakMap w/ frozen keys and garbage collection

### ISSUE THREAD

**Issue by lukescott** [zloirock/core-js#134]
It looks like the current implementation uses a central store when a frozen key is used. This would mean that the keys would never be garbage collected, correct?

I was looking at another WeakMap implementation here: https://github.com/drses/weak-map

It overrides Object.freeze like this (as well as seal and preventExtensions):
```javascript
var oldFreeze = Object.freeze;
Object.freeze = function(obj) {
    Object.defineProperty(obj, HIDDEN_KEY, {
        value: {}
    });
    return oldFreeze(obj);
};
```

This allows it to put a key-store on an object just before it is frozen to give it the opportunity to be used in a weak collection.

This approach may be a bit iffy. The drawbacks are 1) every frozen object gets an object added to it whether it's used or not, and 2) if an object is frozen before the freeze method is wrapped it won't work. But a huge pro to this approach is keys are garbage collected.

**Comment by zloirock** [comment]
Thanks for the proposal. Makes sense, perfect solution.

3) It will not work for `library` version -- global `Object.freeze`-family methods will not be wrapped.

It's required some additional work for prevent side effects, but, I think, I will take it.

**Comment by zloirock** [comment]
4) IE11 with `WeakMap` without support frozen keys and [with separate leaking fix](https://github.com/zloirock/core-js/blob/v1.2.6/modules/es6.weak-map.js#L29-L43).

**Comment by zloirock** [comment]
Available in [2.0](https://github.com/zloirock/core-js/releases/tag/v2.0.0).
